// tasks-client/TaskView.js
var React = require('react.backbone');
var validation = require('backbone-validation');

module.exports = React.createBackboneClass({
  getInitialState: function () {
    return { errors: {} };
  },
  render: require('./Task.jsx'),
  destroy: function (event) {
    console.log("task destroy", this, this.model());

    this.model().destroy();
  },
  save: function (event) {
    console.log("task save", this, this.model());

    var self = this;
    this.model().save(null, {
      success: function () {
        console.log("saved");
        self.setState.call(self, { 'saved': true });
      },
      error: function () {
        console.error("save error");
      },
    });
  },
  onInput: function(prop) {
    console.log("task onInput", this, this.model());
    //this.setState({ 'saved': false });

    var self = this;
    return function () {
      self.model().set(prop, event.target.value);
    };
  },
  componentWillMount: function (rootEl) {
    // bind view to collection
    validation.bind(this, {
      model: this.model(),
    });
  },
});
