// holons-server/Edge.js
"use strict";

/* code ported from https://github.com/DrTom/digraph-demo/blob/master/app/models/arc.rb */

var sql = require('sql');

var Vertex = require('./Vertex');

module.exports = sql.Model.extend({
  tableName: 'holon_arcs',
  initialize: function () {
    this.on('saving', this.preventBackLoop);
    this.on('saving', this.preventCycle);
    this.on('saving', this.preventMultiEdge);
    this.on('saving', this.preventSelfLoop);
  },
  source: function () {
    return this.belongsTo(Vertex, 'sourceId');
  },
  target: function () {
    return this.belongsTo(Vertex, 'targetId');
  },
  successors: function () {
    return this.query('where', 'sourceId', this.get('targetId')).fetch();
  },
  predecessors: function () {
    return this.query('where', 'targetId', this.get('sourceId')).fetch();
  },
  preventBackLoop: function () {
    return;
  },
  preventCycle: function () {
    return;
  },
  preventMultiEdge: function () {
    return;
  },
  preventSelfLoop: function () {
    return;
  },
});
