// holons-server/Vertex.js
"use strict";

/* code ported from https://github.com/DrTom/digraph-demo/blob/master/app/models/node.rb */

var _ = require('underscore');

var sql = require('sql');

var Vertex = module.exports = sql.Model.extend({
  tableName: 'holon_nodes',
  in: function () {
    var Edge = require('./Edge');
    return this.hasMany(Edge, 'targetId');
  },
  out: function () {
    var Edge = require('./Edge');
    return this.hasMany(Edge, 'sourceId');
  },
  parents: function () {
    var Edge = require('./Edge');
    return this.hasMany(Vertex, 'targetId').through(Edge, 'id', 'sourceId');
  },
  children: function () {
    var Edge = require('./Edge');
    return this.hasMany(Vertex, 'sourceId').through(Edge, 'id', 'targetId');
  },
  descendants: function () {
    
  },
  ancestors: function () {

  },
  toJSON: function () {
    var json = sql.Model.prototype.toJSON.call(this);
    // filter out pivot attributes
    return _.pick(json, _.filter(_.keys(json), function (k) { return !k.match(/^_pivot/); }));
  },
});
